Lagrangian particle tracking

ali-ramadhan opened the pull request:
This PR adds Lagrangian particle tracking for incompressible models.

You construct `LagrangianParticles` by passing in a bunch of x, y, z coordinates as 1D arrays (`xs`, `ys`, `zs`) when constructing an incompressible model:

```julia
model = IncompressibleModel(grid = grid, ..., particles = LagrangianParticles(x=xs, y=ys, z=zs))
```

then the particles are advected with the flow at every time step (i.e. online particle advection).

If arrays are used for `x`, `y`, `z` then I think you should be able to add/remove/birth/kill/breed/split/combine/beach/teleport/etc. particles at any time since arrays are mutable and the type signature of the `LagrangianParticles` won't change.

In a futue PR we might want to add `LagrangianParticles` parameters, forcing functions (?), and callafters so that particle properties can be tracked, evolved, and interacted with tracers to model biogeochemistry (https://github.com/JuliaOcean/PlanktonIndividuals.jl) or play ecological games (https://github.com/ali-ramadhan/lagrangian-microbes).

Right now I suppose it only advects particles using forward Euler. Using the model's RK3 for particle advection would double (or triple?) the memory usage of the particle tracker as we would have to store previous tendencies (velocities). I wonder if the accuracy of forward Euler is acceptable for Lagrangian particle tracking. I'm also wondering whether it makes sense to support both forward Euler and RK3 in case you want to advect a huge number of particles and memory is limiting the amount of particles you can track? I guess you can do RK3 or RK4 inside the `advect_particles!` kernel but it would assume that the velocity field is constant between stages. Perhaps a convergence test is needed to answer these questions.

I think there are a lot of things that can be done at boundaries (see e.g. https://github.com/OceanParcels/parcels/issues/37, https://github.com/OceanParcels/parcels/issues/47, https://github.com/OceanParcels/parcels/issues/61) so I think we should pick a decent default and leave other options for future PRs.

Right now, periodic dimensions are handled correctly but otherwise if a particle goes through a wall it will be placed back at the wall. This may cause it to get stuck at the wall.

cc @suyash @zhenwu0728 Would something like this work for you guys?

TODO:

- [x] Figure out how we want to time step particle advection.
- [x] Decide on a default way of handling boundary conditions.
- [x] Tell NetCDF and JLD2 output writers how to write `LagrangianParticles` locations to disk.
- [x] Test that particle locations are correct after advection. Can test with an analytic solution produced by a constant flow field.
- [x] Write documentation for how to use Lagrangian particle tracking (add a page to the model setup docs).
- [x] Benchmark Lagrangian particle tracking with 1 - 100,000,000 particles.
- [x] Make cool movie.

Resolves #511

ali-ramadhan commented on src/LagrangianParticleTracking.jl:
```diff
@@ -0,0 +1,84 @@
+
+@inline enforce_periodicity(x, xₗ, xᵣ, ::Type{Bounded})  = x
+
+@inline function enforce_periodicity(x, xₗ, xᵣ, ::Type{Periodic})
+    x > xᵣ && return xₗ + (x - xᵣ)
+    x < xₗ && return xᵣ - (xₗ - x)
+    return x
+end
```
Just started with a more readable function but couldn't figure out how to rewrite using e.g. `mod` functions to get rid of the conditional statements.

zhenwu0728 replied:
Hi Ali - currently I am using the code below for periodic domain, which is easy to broadcast. Not sure if this suits your situation.
Also, in `PlanktonIndividuals.jl` I use array broadcast instead of `KernelAbstractions.jl` for partical advection on GPU. You can see my code on `dev_gpu` branch.
```julia
x = isless(x, xᵣ) * x + isequal(x, xᵣ) * x + isless(xᵣ, x) * (x - xᵣ + xₗ)
x = isless(xₗ, x) * x + isequal(xₗ, x) * x + isless(x, xₗ) * (xᵣ - xₗ + x )
```

ali-ramadhan replied:
Thanks @zhenwu0728 that's pretty neat! Didn't know about `isless` and `isequal`.

ali-ramadhan replied:
Lots of work on your `dev_gpu` branch! Out of curiosity how many particles/plankton/microbes do you usually simulate? Is it a big performance hit?

Maybe I should open an issue on PlanktonIndividuals.jl to ask.

zhenwu0728 replied:
That's OK. I usually use 1,000-10,000 particles. But I ran them on the CPU. It takes hours to run.
The GPU part is still working in progress. So I cannot tell you the benchmarks on GPU right now.  
A rough comparison was at least 3 times faster on GPU when I was testing my code.

glwagner commented on src/LagrangianParticleTracking.jl:
```diff
@@ -0,0 +1,84 @@
+    x :: X
+    y :: Y
+    z :: Z
+end
```
I haven't worked out all the details, but I wonder if abstracting `model.particles` as an `Array` or `CuArray` of `AbstractParticle`s is a better design. Consider:

```julia
struct ReactingTracerParticle{P} <: AbstractParticle
    position :: P
    species :: P
end
```

We can then provide an interface for defining reactions and such things on particles.

This use is futuristic, but I don't think we lose anything by modifying the design to anticipate uses like this. It's also simpler because we don't have to check that `x, y, z` are the same length, etc.

glwagner replied:
Hmm, perhaps what we would want to support is a type that can extract field values at the particle position.

Then we can design an interface for writing RHS that depend on the field values so extracted, so that reactions-on-particles can depend arbitrarily on tracer/velocity values at the particle location.

This more advanced type doesn't have to be part of this PR but I think we should set off on the right foot.

zhenwu0728 replied:
Not sure you know about `StructArrays.jl`. I'm using it now for particles in `PlanktonIndividuals.jl`. It works on both CPU and GPU.

glwagner replied:
That's interesting --- but actually wouldn't it be more efficient to keep each position `x, y, z` close in memory? For example, @ali-ramadhan's kernel iterates over the list of particles, rather than each coordinate individually. This is better served by the layout

`[(x1, y1, z1), (x2, y1, z2), ...]`

(though in practice it may not matter much).

glwagner replied:
`StructArrays.jl` could definitely prove useful in some capacity so thanks for pointing it out!

zhenwu0728 replied:
Yes, you're right. The reason I use `StructArrays.jl` is that it is easy to broadcast at `x, y, z`. But if you use `@kernel` functions, the first one is more readable.

glwagner commented on src/LagrangianParticleTracking.jl:
```diff
@@ -0,0 +1,84 @@
+    # If particles go through a wall, put them back at the wall.
+    @inbounds particles.x[p] = clamp(particles.x[p], grid.xF[1], grid.xF[grid.Nx])
+    @inbounds particles.y[p] = clamp(particles.y[p], grid.yF[1], grid.yF[grid.Ny])
+    @inbounds particles.z[p] = clamp(particles.z[p], grid.zF[1], grid.zF[grid.Nz])
```
How about just one function called `enforce_boundary_conditions` that dispatches on `TX`?

Imagine

```julia
struct BouncingParticle{P, R} <: AbstractParticle
    position :: P
    restitution :: R
end
```

Then `restitution = 0` we get "clamping" behavior; if `restitution = 1` we get perfect reflection.

glwagner replied:
Or if `restitution` is a 3-tuple, we get different behavior at the 3 walls.

glwagner commented:
I left a few comments in the code. To guide the design of this feature, consider a few important use cases:

1. Buoyant particles (eg particles that feel an "additional" advecting velocity, in addition to `model.velocities` and `model.background_fields.velocities`)
2. Chemical reactions taking place on particles
3. "Elastic" particles that bounce of walls with a specified coefficient of restitution
4. Particles that are nudged with random stochastic noise in addition to advection (this is a crucial feature that is necessary to approximate the effect of molecular / subfilter turbulent diffusion)

I am not suggesting that we support all of these, but it'd be good to have a type-based infrastructure for particle dynamics that can be extended to include more features in the future, in particular without breaking the existing API and rewriting old code.

glwagner commented:
Perhaps a solution is to design a `Particle` that looks something like

If we have a `Particle` design that looks something like

```julia
struct Particle{P, T, R, E, D}
    position :: P
    properties :: T
    restitution :: R
    external_velocities :: E
    diffusivity :: D
end
```

With the following meaning:

* `position`: a 3-element object with the particle position (not sure what's best for GPU --- custom type or `Array` / `CuArray`?
* `properties` (could use a better name): user-defined properties + model fields for simulating reactions and such. Probably need additional types / interface for these
* `restitution`: floating point number between 0 and 1 that determines particle behavior at solid walls
* `external_velocities`: a 3-tuple of settling/buoyant/other velocities to be applied to the particle in addition to the resolved + background velocity field. We can also design an abstraction / special type that computes the buoyancy of a particle with a certain temperature / salinity / buoyancy (for example)
* `diffusivity` diffusion coefficient that is used to compute stochastic noise added to the particle position (requires some care to ensure the stochastic diff eq is integrated correctly. Would be nice to support both constants and functions of space-time, etc.

We can start with just a few of these fields in our `Particle` type, and extend its power in future PRs if we need?

christophernhill commented:
@ali-ramadhan and @glwagner adding @renskegelderloos to this PR so she can meet you. 

Renske is interested in the world of Ocean particle tools in Julia, timesteppers, GPUs etc....

suyashbire1 commented on src/LagrangianParticleTracking.jl:
```diff
@@ -0,0 +1,84 @@
+        error("x, y, z must all have the same size!")
+        
+    (ndims(x) == 1 && ndims(y) == 1 && ndims(z) == 1) ||
+        error("x, y, z must have dimension 1 but ndims=($(ndims(x)), $(ndims(y)), $(ndims(z)))")
```
Also check if each particle's initial position is inside the domain.

ali-ramadhan commented:
Hi @renskegelderloos! What sort of stuff are you working on? This is definitely a pretty barebones Lagrangian particle tracker right now, but hoping to get it working on the GPU soon!

Sorry been meaning to get back to this PR!

renskegelderloos commented:
Hi @ali-ramadhan! I'm working on Lagrangian analysis tools for ocean model output. Our original code was written in Matlab (https://doi.org/10.1109/escience.2016.7870923), but since I've recently been learning more about Julia I would like to switch. Our goal is to provide easy community access to large data volume numerical ocean simulations, with ready-made and easy-to-use tools. For Eulerian analysis we've used Python (https://doi.org/10.21105/joss.01506); that package is up and running. I want to add Lagrangian analysis tools to this framework. I'd love to learn more about what you guys are up to!

ali-ramadhan commented:
@renskegelderloos Thanks for the paper! Not sure if it was used for your [particle tracking example notebook](https://oceanspy.readthedocs.io/en/latest/Particles.html) but looks pretty neat!

We could chat over Slack or Zoom with @christophernhill @glwagner + anyone else interested?

Cool to see GPU benchmarks in your paper, would love to compare there. Might even make sense to implement your algorithm in Julia? Are you thinking of implementing GPU particle tracking in Python/Numba?

I guess we're doing online particle advection where time steps are pretty small and the velocities are available at every time step, so maybe forward Euler + trilinear interpolation is fine but maybe not?

ali-ramadhan commented:
Some pretty promising Lagrangian particle tracking benchmarks!

Couple of takeaways (all assuming a model with 128^3 grid points and QAB2 time stepping):
1. **Low overhead**: You can advect up to ~100,000 particles on the CPU and up to ~10,000,000 particles on a (Titan V) GPU before the model slows down by more than 30%.
2. **Great on GPUs**: Seems that the GPU is great for advecting millions of particles. You can advect ~100,000,000 particles and your model only slows down by a factor of 4x. In this scenario, the GPU is ~620x faster than a single CPU core.
3. Calculated using `(t_100000000 - t_0) / 100000000`, advecting a single particle on the CPU takes ~110 ns while on the GPU it only takes ~0.127 ns. This seems a little too good to be true but I'll double check this.

I'll start refactoring this PR using @glwagner's and @zhenwu0728's feedback, but I think it would be really great if we can keep this performance.

# Benchmarks

```
Oceananigans v0.44.1
Julia Version 1.5.2
Commit 539f3ce943 (2020-09-23 23:17 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Silver 4214 CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, cascadelake)
  GPU: TITAN V
```

```
                                 Lagrangian particle tracking benchmarks
┌───────────────┬─────────────┬────────────┬────────────┬────────────┬────────────┬────────────┬────────┐
│ Architectures │ N_particles │        min │     median │       mean │        max │     memory │ allocs │
├───────────────┼─────────────┼────────────┼────────────┼────────────┼────────────┼────────────┼────────┤
│           CPU │           0 │ 361.749 ms │ 364.041 ms │ 364.293 ms │ 368.854 ms │ 293.44 KiB │   1876 │
│           CPU │           1 │ 375.030 ms │ 376.591 ms │ 377.959 ms │ 385.248 ms │ 297.16 KiB │   1906 │
│           CPU │          10 │ 377.251 ms │ 380.792 ms │ 387.560 ms │ 443.325 ms │ 297.16 KiB │   1906 │
│           CPU │         100 │ 378.867 ms │ 381.194 ms │ 381.328 ms │ 383.461 ms │ 297.16 KiB │   1906 │
│           CPU │        1000 │ 378.076 ms │ 384.114 ms │ 383.611 ms │ 388.507 ms │ 297.16 KiB │   1906 │
│           CPU │       10000 │ 381.209 ms │ 383.687 ms │ 387.584 ms │ 419.196 ms │ 297.16 KiB │   1906 │
│           CPU │      100000 │ 389.478 ms │ 390.299 ms │ 390.442 ms │ 391.984 ms │ 297.16 KiB │   1906 │
│           CPU │     1000000 │ 484.920 ms │ 487.836 ms │ 488.382 ms │ 493.664 ms │ 297.16 KiB │   1906 │
│           CPU │    10000000 │    1.484 s │    1.489 s │    1.497 s │    1.525 s │ 297.16 KiB │   1906 │
│           CPU │   100000000 │   11.523 s │   11.523 s │   11.523 s │   11.523 s │ 297.16 KiB │   1906 │
│           GPU │           0 │   4.410 ms │   4.427 ms │   4.505 ms │   5.169 ms │ 884.91 KiB │  11409 │
│           GPU │           1 │   4.480 ms │   4.527 ms │   4.584 ms │   4.871 ms │ 897.25 KiB │  11563 │
│           GPU │          10 │   4.511 ms │   4.550 ms │   4.662 ms │   5.615 ms │ 896.53 KiB │  11517 │
│           GPU │         100 │   4.515 ms │   4.531 ms │   4.645 ms │   5.490 ms │ 896.53 KiB │  11517 │
│           GPU │        1000 │   4.484 ms │   4.540 ms │   4.593 ms │   5.115 ms │ 897.25 KiB │  11563 │
│           GPU │       10000 │   4.517 ms │   4.564 ms │   4.687 ms │   5.695 ms │ 896.53 KiB │  11517 │
│           GPU │      100000 │   4.471 ms │   4.530 ms │   4.631 ms │   5.623 ms │ 896.53 KiB │  11517 │
│           GPU │     1000000 │   4.520 ms │   4.584 ms │   4.703 ms │   5.666 ms │ 897.25 KiB │  11563 │
│           GPU │    10000000 │   5.709 ms │   5.842 ms │   5.836 ms │   5.917 ms │ 897.22 KiB │  11561 │
│           GPU │   100000000 │   4.910 ms │  18.629 ms │  17.245 ms │  18.757 ms │ 896.56 KiB │  11519 │
└───────────────┴─────────────┴────────────┴────────────┴────────────┴────────────┴────────────┴────────┘
```

```
   Lagrangian particle tracking CPU -> GPU
                   speedup
┌─────────────┬─────────┬─────────┬─────────┐
│ N_particles │ speedup │  memory │  allocs │
├─────────────┼─────────┼─────────┼─────────┤
│           0 │ 82.2319 │ 3.01565 │ 6.08156 │
│           1 │ 83.1848 │ 3.01946 │ 6.06663 │
│          10 │ 83.6845 │ 3.01704 │  6.0425 │
│         100 │ 84.1322 │ 3.01704 │  6.0425 │
│        1000 │ 84.6106 │ 3.01946 │ 6.06663 │
│       10000 │  84.072 │ 3.01704 │  6.0425 │
│      100000 │ 86.1581 │ 3.01704 │  6.0425 │
│     1000000 │ 106.432 │ 3.01946 │ 6.06663 │
│    10000000 │ 254.889 │ 3.01935 │ 6.06558 │
│   100000000 │ 618.565 │ 3.01714 │ 6.04355 │
└─────────────┴─────────┴─────────┴─────────┘
```

```
   Lagrangian particle tracking relative performance (CPU)
┌───────────────┬─────────────┬──────────┬─────────┬─────────┐
│ Architectures │ N_particles │ slowdown │  memory │  allocs │
├───────────────┼─────────────┼──────────┼─────────┼─────────┤
│           CPU │           0 │      1.0 │     1.0 │     1.0 │
│           CPU │           1 │  1.03447 │ 1.01267 │ 1.01599 │
│           CPU │          10 │  1.04601 │ 1.01267 │ 1.01599 │
│           CPU │         100 │  1.04712 │ 1.01267 │ 1.01599 │
│           CPU │        1000 │  1.05514 │ 1.01267 │ 1.01599 │
│           CPU │       10000 │  1.05397 │ 1.01267 │ 1.01599 │
│           CPU │      100000 │  1.07213 │ 1.01267 │ 1.01599 │
│           CPU │     1000000 │  1.34006 │ 1.01267 │ 1.01599 │
│           CPU │    10000000 │  4.09045 │ 1.01267 │ 1.01599 │
│           CPU │   100000000 │  31.6534 │ 1.01267 │ 1.01599 │
└───────────────┴─────────────┴──────────┴─────────┴─────────┘
```

```
   Lagrangian particle tracking relative performance (GPU)
┌───────────────┬─────────────┬──────────┬─────────┬─────────┐
│ Architectures │ N_particles │ slowdown │  memory │  allocs │
├───────────────┼─────────────┼──────────┼─────────┼─────────┤
│           GPU │           0 │      1.0 │     1.0 │     1.0 │
│           GPU │           1 │  1.02262 │ 1.01395 │  1.0135 │
│           GPU │          10 │  1.02786 │ 1.01314 │ 1.00947 │
│           GPU │         100 │  1.02347 │ 1.01314 │ 1.00947 │
│           GPU │        1000 │  1.02548 │ 1.01395 │  1.0135 │
│           GPU │       10000 │   1.0309 │ 1.01314 │ 1.00947 │
│           GPU │      100000 │  1.02327 │ 1.01314 │ 1.00947 │
│           GPU │     1000000 │  1.03536 │ 1.01395 │  1.0135 │
│           GPU │    10000000 │  1.31966 │ 1.01391 │ 1.01332 │
│           GPU │   100000000 │  4.20799 │ 1.01317 │ 1.00964 │
└───────────────┴─────────────┴──────────┴─────────┴─────────┘
```

johncmarshall54 commented:
Let's go for 100million then! Great stats Ali. Can we put them in to the
icy moon runs to get pdf of transit time from bottom to top. We need a
clock on each one and we stop it when the particle gets closer than a
chosen distance from the upper boundary. John

On Mon, Nov 23, 2020, 8:35 PM Ali Ramadhan <notifications@github.com> wrote:

> Some pretty promising Lagrangian particle tracking benchmarks!
>
> Couple of takeaways (all assuming a model with 128^3 grid points and QAB2
> time stepping):
>
>    1. *Low overhead*: You can advect up to ~100,000 particles on the CPU
>    and up to ~10,000,000 particles on a (Titan V) GPU before the model slows
>    down by more than 30%.
>    2. *Great on GPUs*: Seems that the GPU is great for advecting millions
>    of particles. You can advect ~100,000,000 particles and your model only
>    slows down by a factor of 4x. In this scenario, the GPU is ~620x faster
>    than a single CPU core.
>    3. Calculated using (t_100000000 - t_0) / 100000000, advecting a
>    single particle on the CPU takes ~110 ns while on the GPU it only takes
>    ~0.127 ns. This seems a little too good to be true but I'll double check
>    this.
>
> I'll start refactoring this PR using @glwagner
> <https://github.com/glwagner>'s and @zhenwu0728
> <https://github.com/zhenwu0728>'s feedback, but I think it would be
> really great if we can keep this performance.
> Benchmarks
>
> Oceananigans v0.44.1
> Julia Version 1.5.2
> Commit 539f3ce943 (2020-09-23 23:17 UTC)
> Platform Info:
>   OS: Linux (x86_64-pc-linux-gnu)
>   CPU: Intel(R) Xeon(R) Silver 4214 CPU @ 2.20GHz
>   WORD_SIZE: 64
>   LIBM: libopenlibm
>   LLVM: libLLVM-9.0.1 (ORCJIT, cascadelake)
>   GPU: TITAN V
>
>                                  Lagrangian particle tracking benchmarks
> ┌───────────────┬─────────────┬────────────┬────────────┬────────────┬────────────┬────────────┬────────┐
> │ Architectures │ N_particles │        min │     median │       mean │        max │     memory │ allocs │
> ├───────────────┼─────────────┼────────────┼────────────┼────────────┼────────────┼────────────┼────────┤
> │           CPU │           0 │ 361.749 ms │ 364.041 ms │ 364.293 ms │ 368.854 ms │ 293.44 KiB │   1876 │
> │           CPU │           1 │ 375.030 ms │ 376.591 ms │ 377.959 ms │ 385.248 ms │ 297.16 KiB │   1906 │
> │           CPU │          10 │ 377.251 ms │ 380.792 ms │ 387.560 ms │ 443.325 ms │ 297.16 KiB │   1906 │
> │           CPU │         100 │ 378.867 ms │ 381.194 ms │ 381.328 ms │ 383.461 ms │ 297.16 KiB │   1906 │
> │           CPU │        1000 │ 378.076 ms │ 384.114 ms │ 383.611 ms │ 388.507 ms │ 297.16 KiB │   1906 │
> │           CPU │       10000 │ 381.209 ms │ 383.687 ms │ 387.584 ms │ 419.196 ms │ 297.16 KiB │   1906 │
> │           CPU │      100000 │ 389.478 ms │ 390.299 ms │ 390.442 ms │ 391.984 ms │ 297.16 KiB │   1906 │
> │           CPU │     1000000 │ 484.920 ms │ 487.836 ms │ 488.382 ms │ 493.664 ms │ 297.16 KiB │   1906 │
> │           CPU │    10000000 │    1.484 s │    1.489 s │    1.497 s │    1.525 s │ 297.16 KiB │   1906 │
> │           CPU │   100000000 │   11.523 s │   11.523 s │   11.523 s │   11.523 s │ 297.16 KiB │   1906 │
> │           GPU │           0 │   4.410 ms │   4.427 ms │   4.505 ms │   5.169 ms │ 884.91 KiB │  11409 │
> │           GPU │           1 │   4.480 ms │   4.527 ms │   4.584 ms │   4.871 ms │ 897.25 KiB │  11563 │
> │           GPU │          10 │   4.511 ms │   4.550 ms │   4.662 ms │   5.615 ms │ 896.53 KiB │  11517 │
> │           GPU │         100 │   4.515 ms │   4.531 ms │   4.645 ms │   5.490 ms │ 896.53 KiB │  11517 │
> │           GPU │        1000 │   4.484 ms │   4.540 ms │   4.593 ms │   5.115 ms │ 897.25 KiB │  11563 │
> │           GPU │       10000 │   4.517 ms │   4.564 ms │   4.687 ms │   5.695 ms │ 896.53 KiB │  11517 │
> │           GPU │      100000 │   4.471 ms │   4.530 ms │   4.631 ms │   5.623 ms │ 896.53 KiB │  11517 │
> │           GPU │     1000000 │   4.520 ms │   4.584 ms │   4.703 ms │   5.666 ms │ 897.25 KiB │  11563 │
> │           GPU │    10000000 │   5.709 ms │   5.842 ms │   5.836 ms │   5.917 ms │ 897.22 KiB │  11561 │
> │           GPU │   100000000 │   4.910 ms │  18.629 ms │  17.245 ms │  18.757 ms │ 896.56 KiB │  11519 │
> └───────────────┴─────────────┴────────────┴────────────┴────────────┴────────────┴────────────┴────────┘
>
>    Lagrangian particle tracking CPU -> GPU
>                    speedup
> ┌─────────────┬─────────┬─────────┬─────────┐
> │ N_particles │ speedup │  memory │  allocs │
> ├─────────────┼─────────┼─────────┼─────────┤
> │           0 │ 82.2319 │ 3.01565 │ 6.08156 │
> │           1 │ 83.1848 │ 3.01946 │ 6.06663 │
> │          10 │ 83.6845 │ 3.01704 │  6.0425 │
> │         100 │ 84.1322 │ 3.01704 │  6.0425 │
> │        1000 │ 84.6106 │ 3.01946 │ 6.06663 │
> │       10000 │  84.072 │ 3.01704 │  6.0425 │
> │      100000 │ 86.1581 │ 3.01704 │  6.0425 │
> │     1000000 │ 106.432 │ 3.01946 │ 6.06663 │
> │    10000000 │ 254.889 │ 3.01935 │ 6.06558 │
> │   100000000 │ 618.565 │ 3.01714 │ 6.04355 │
> └─────────────┴─────────┴─────────┴─────────┘
>
>    Lagrangian particle tracking relative performance (CPU)
> ┌───────────────┬─────────────┬──────────┬─────────┬─────────┐
> │ Architectures │ N_particles │ slowdown │  memory │  allocs │
> ├───────────────┼─────────────┼──────────┼─────────┼─────────┤
> │           CPU │           0 │      1.0 │     1.0 │     1.0 │
> │           CPU │           1 │  1.03447 │ 1.01267 │ 1.01599 │
> │           CPU │          10 │  1.04601 │ 1.01267 │ 1.01599 │
> │           CPU │         100 │  1.04712 │ 1.01267 │ 1.01599 │
> │           CPU │        1000 │  1.05514 │ 1.01267 │ 1.01599 │
> │           CPU │       10000 │  1.05397 │ 1.01267 │ 1.01599 │
> │           CPU │      100000 │  1.07213 │ 1.01267 │ 1.01599 │
> │           CPU │     1000000 │  1.34006 │ 1.01267 │ 1.01599 │
> │           CPU │    10000000 │  4.09045 │ 1.01267 │ 1.01599 │
> │           CPU │   100000000 │  31.6534 │ 1.01267 │ 1.01599 │
> └───────────────┴─────────────┴──────────┴─────────┴─────────┘
>
>    Lagrangian particle tracking relative performance (GPU)
> ┌───────────────┬─────────────┬──────────┬─────────┬─────────┐
> │ Architectures │ N_particles │ slowdown │  memory │  allocs │
> ├───────────────┼─────────────┼──────────┼─────────┼─────────┤
> │           GPU │           0 │      1.0 │     1.0 │     1.0 │
> │           GPU │           1 │  1.02262 │ 1.01395 │  1.0135 │
> │           GPU │          10 │  1.02786 │ 1.01314 │ 1.00947 │
> │           GPU │         100 │  1.02347 │ 1.01314 │ 1.00947 │
> │           GPU │        1000 │  1.02548 │ 1.01395 │  1.0135 │
> │           GPU │       10000 │   1.0309 │ 1.01314 │ 1.00947 │
> │           GPU │      100000 │  1.02327 │ 1.01314 │ 1.00947 │
> │           GPU │     1000000 │  1.03536 │ 1.01395 │  1.0135 │
> │           GPU │    10000000 │  1.31966 │ 1.01391 │ 1.01332 │
> │           GPU │   100000000 │  4.20799 │ 1.01317 │ 1.00964 │
> └───────────────┴─────────────┴──────────┴─────────┴─────────┘
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CliMA/Oceananigans.jl/pull/1091#issuecomment-732529975>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKXUEQS3WD4CLJSVF23H433SRMEWVANCNFSM4SZUOFJQ>
> .
>

glwagner commented:
@ali-ramadhan agree that good performance would be great. I think that having a flexible and maximally useful feature might be a higher priority though. @johncmarshall54 's requirements provide a good design target, I think.

ali-ramadhan commented:
I agree that having a flexible particle tracking framework is more important. I've already started refactoring with @suyashbire1.

Just wanted to benchmark the barebones implementation first to ensure that the extra features don't slow it down by 100x. Sometimes it's easy to introduce huge slowdowns on the GPU.

ali-ramadhan commented:
Good news is that using a `StructArray{Particle}` doesn't impact performance :rocket:

ali-ramadhan commented:
One annoying thing about using `StructArray{Particle}` is that the array of particle is now stored in `model.particles.particles`.

I guess `model.lagrangian_particles.particles` or `model.particles.particle_set` are potential solutions.

```julia
struct Particle{T} <: AbstractParticle
    x :: T
    y :: T
    z :: T
end

# This struct will hold more properties in the future.
struct LagrangianParticles{P}
    particles :: P
end
```

zhenwu0728 commented:
> ```julia
> struct Particle{T} <: AbstractParticle
>     x :: T
>     y :: T
>     z :: T
> end
> ```
In PlanktonIndividuals.jl, I'm using a `StructArray` like this:
```julia
struct plankton{T}
x::T
y::T
z::T
cell_size::T
Biomass::T
age::T
...
end
```

> ```julia
> # This struct will hold more properties in the future.
> struct LagrangianParticles{P}
>     particles :: P
> end
> ```
Yes, but need to be careful when you remove some particles from the particle set. You need to remove the particle and other properties at the same time, otherwise the particles and properties will not match.

ali-ramadhan commented:
Ah thanks for pointing this out @zhenwu0728! I didn't realize that.

It also seems that `deleteat!` is not defined for `StructArray`...

```julia
julia> s = StructArray{ComplexF64}((randn(5), randn(5)))
5-element StructArray(::Array{Float64,1}, ::Array{Float64,1}) with eltype Complex{Float64}:
  0.6663581952584336 - 1.5283519296646326im
 -0.4906568589874505 - 0.325094980686686im
  1.1453745422319603 + 0.07874273417016024im
  1.0280155738262986 - 0.37561698848991654im
  1.1466416201897343 - 1.1904623364760012im

julia> deleteat!(s, 2)
ERROR: MethodError: no method matching deleteat!(::StructArray{Complex{Float64},1,NamedTuple{(:re, :im),Tuple{Array{Float64,1},Array{Float64,1}}},Int64}, ::Int64)
Closest candidates are:
  deleteat!(::Array{T,1} where T, ::Integer) at array.jl:1306
  deleteat!(::Array{T,1} where T, ::Any) at array.jl:1343
  deleteat!(::BitArray{1}, ::Integer) at bitarray.jl:935
  ...
Stacktrace:
 [1] top-level scope at REPL[13]:1
```

@suyashbire1 and I decided to focus on just supporting a constant number of particles for now since it's not super clear what the best approach is for saving particle locations to disk when the number of particles keeps changing.

zhenwu0728 commented:
> Ah thanks for pointing this out @zhenwu0728! I didn't realize that.
> 
> It also seems that `deleteat!` is not defined for `StructArray`...
> 
> ```julia
> julia> s = StructArray{ComplexF64}((randn(5), randn(5)))
> 5-element StructArray(::Array{Float64,1}, ::Array{Float64,1}) with eltype Complex{Float64}:
>   0.6663581952584336 - 1.5283519296646326im
>  -0.4906568589874505 - 0.325094980686686im
>   1.1453745422319603 + 0.07874273417016024im
>   1.0280155738262986 - 0.37561698848991654im
>   1.1466416201897343 - 1.1904623364760012im
> 
> julia> deleteat!(s, 2)
> ERROR: MethodError: no method matching deleteat!(::StructArray{Complex{Float64},1,NamedTuple{(:re, :im),Tuple{Array{Float64,1},Array{Float64,1}}},Int64}, ::Int64)
> Closest candidates are:
>   deleteat!(::Array{T,1} where T, ::Integer) at array.jl:1306
>   deleteat!(::Array{T,1} where T, ::Any) at array.jl:1343
>   deleteat!(::BitArray{1}, ::Integer) at bitarray.jl:935
>   ...
> Stacktrace:
>  [1] top-level scope at REPL[13]:1
> ```

Yes, I think so. But there's `StructArrays.append!!` to add particles. I don't know why there's no `delete!!` or `deleteat!!`. Maybe we could open an issue in `StructArrays.jl`. Also we need to pay attention to the performance as well.

ali-ramadhan commented:
@zhenwu0728 Looks like there's an open issue about this: https://github.com/JuliaArrays/StructArrays.jl/issues/98

Apparently one solution is

```julia
foreachfield(v -> deleteat!(v, i), s)
```

from https://github.com/JuliaArrays/StructArrays.jl/issues/144#issuecomment-669892119

zhenwu0728 commented:
> @zhenwu0728 Looks like there's an open issue about this: [JuliaArrays/StructArrays.jl#98](https://github.com/JuliaArrays/StructArrays.jl/issues/98)
> 
> Apparently one solution is
> 
> ```julia
> foreachfield(v -> deleteat!(v, i), s)
> ```
> 
> from [JuliaArrays/StructArrays.jl#144 (comment)](https://github.com/JuliaArrays/StructArrays.jl/issues/144#issuecomment-669892119)

Thanks @ali-ramadhan! I should try this out. 
But what I'm worried about is that the kernel functions need to recompile if the array size changed. Is there a solution to that?

ali-ramadhan commented:
> But what I'm worried about is that the kernel functions need to recompile if the array size changed. Is there a solution to that?

Ah I think this should be fine as long as kernels are launched with the correct work size. Might only recompile if the kernel dispatches on the number of particles but that's only possible with StaticArrays.jl I think (`StructArray` stores the number of dimensions in the type but not the number of elements/particles).

I tried to look through your kernels in https://github.com/JuliaOcean/PlanktonIndividuals.jl/pull/17 but I think you should be fine since you pass the number of particles in when launching kernels, e.g. https://github.com/JuliaOcean/PlanktonIndividuals.jl/pull/17/files#diff-effe03966dbceff3dc287bb5dd64df44fa42c03af8d9f569b951a6ca1e0133d3R30

zhenwu0728 commented:
> Ah I think this should be fine as long as kernels are launched with the correct work size. Might only recompile if the kernel dispatches on the number of particles but that's only possible with StaticArrays.jl I think (`StructArray` stores the number of dimensions in the type but not the number of elements/particles). 

Thanks for letting me know. That's definitely good news.

ali-ramadhan commented:
I think this PR is ready for another review @glwagner.

Users can now specify their own particle types with custom properties, e.g.

```julia
struct Microbe{T, S, D}
    x :: T
    y :: T
    z :: T
    w :: T
    T :: T
    species :: S
    dna :: D
end
```

where `w` and `T` can store the w and T at the particle's location if `tracked_fields = (w=model.velocities.w, T=model.tracers.T)` is passed to `LagrangianParticles`. Tracked fields can include `ComputedField`s (see the tests).

Custom properties, e.g. `species` and `dna`, can be modified via a callback.

We might have to figure out callbacks for `LagrangianParticles`. I think ideally we'll want a schedule for the callback so might make sense to do this with/after #1138. Right now it's kind of a "hidden feature" as I modify custom particle properties as part of the simulation callback.

I should run benchmarks for 0-10 tracked fields.

glwagner commented on src/LagrangianParticleTracking/LagrangianParticleTracking.jl:
```diff
@@ -0,0 +1,66 @@
+         particles :: P
+       restitution :: R
+    tracked_fields :: T
+end
```
For the sake of future dev's and myself and also the docs, do you mind providing some comments / docs on the purpose of the `LagrangianParticles` struct? Naively I would think it would make sense to store everything we need particle-wise so that `model.particles` would just be a simple `Array` / `CuArray`. I'm sure there's a good reason for this decision, just can't detect it at a glance.

ali-ramadhan replied:
I added some model setup docs for Lagrangian particle tracking showing how to track simple particles and particles with custom properties (+ writing particle properties to JLD2/NetCDF).

As for the purpose of the `LagrangianParticles` struct: Initially I just kept track of the particle's locations/properties so a single `StructArray` was enough. But then it became obvious that there would be some properties shared by all particles, e.g. a coefficient of restitution and which fields are being tracked. There may be more in the future if we add more features.

The coefficient of restitution could be stored by each particle but that would increase memory requirements for the narrow case of wanting a different coefficient for each particle.

glwagner replied:
> But then it became obvious that there would be some properties shared by all particles, e.g. a coefficient of restitution and which fields are being tracked.

Why do we want to enforce that all particles have the same coefficient of restitution? Isn't this a limitation?

glwagner replied:
I would advocate for an implementation that's as flexible as possible, even if it does impose some extra memory or computational requirements. I think the new science afforded by a flexible abstraction might be more important than faster simulations or the ability to run with huge amounts of particles, right? We can always provide "optimized" implementations if that becomes important in the future; but we might not want to limit what people can do for performance whose necessity is unclear?

ali-ramadhan replied:
The current implementation is flexible enough to do almost anything.

You can assign different coefficients of restitution to each particle and implement any kind of particle-wall collision you want.

See #1244 for an example of using custom particle types to implement particle-particle interactions.

But yes in this PR by default there is one coefficient of restitution for all particles. If you think it's important that each particle has it's own coefficient I'm happy to make the change if it means the PR can be merged.

glwagner commented on src/LagrangianParticleTracking/advect_particles.jl:
```diff
@@ -0,0 +1,78 @@
+    x > xᵣ && return xᵣ - (x - xᵣ) * restitution
+    x < xₗ && return xₗ + (xₗ - x) * restitution
+    return x
+end
```
🌹

glwagner commented on src/LagrangianParticleTracking/advect_particles.jl:
```diff
@@ -0,0 +1,78 @@
+    # Advect particles using forward Euler.
+    @inbounds particles.x[p] += interpolate(velocities.u, Face(), Cell(), Cell(), grid, particles.x[p], particles.y[p], particles.z[p]) * Δt
+    @inbounds particles.y[p] += interpolate(velocities.v, Cell(), Face(), Cell(), grid, particles.x[p], particles.y[p], particles.z[p]) * Δt
+    @inbounds particles.z[p] += interpolate(velocities.w, Cell(), Cell(), Face(), grid, particles.x[p], particles.y[p], particles.z[p]) * Δt
```
How does this work given that `Particles` is not `mutable` (am I confuzzled?)

ali-ramadhan replied:
Ah it all works since particle locations/properties are stored in a `StructArray`. `particles.x` is an array but `particles[1]` reconstructs a `Particle` and returns it.

Not completely sure how it works but I think it converts an "array of structs" into "arrays for each property" so it becomes easy to pull out all the x locations to save particle locations to disk. It also becomes easy to add/remove particles.

ali-ramadhan commented:
@glwagner Sorry for leaving this PR open for so long. I added some model setup docs and added tests for JLD2 particle output so I think it should be ready to be merged now.

I also bumped v0.46.0 in this PR.

ali-ramadhan commented:
@glwagner Made the changes so I'll finally merge now.
